### ISSUE TITLE
Process on getting test data needed for creating test cases

### ISSUE THREAD

**Issue by veydecapia** [paritytech/substrate-api-sidecar#808]
I know this is not an issue but if someone can give me an idea on the process of getting the test data needed for building test cases, that would be greatly appreciated.

So I'm trying to create an end to end test for Karura and I'm modeling the same boilerplate code on other networks. 
And my concern is **how does one get a test data (account addresses, block height, etc) that will represent the endpoint string for different test cases (staking, vesting, crowdloan, lease, auction, paras, etc.)?**
Tried the subscan, https://karura.subscan.io/event for getting account addresses and block number.
I did get some account addresses needed for balanceinfo test cases but not so much success for other tests like staking,vesting.

I've already created and done some test cases of my own based on the already created test cases.
https://github.com/veydecapia/substrate-api-sidecar/tree/harvey-add-karura-test/e2e-tests/endpoints/karura

Thanks in advance!

**Comment by TarikGul** [comment]
Hey @veydecapia!

As a helper PR too, you can check #807 as a reference as that PR is adding `Statemint` as a  new chain to the e2e tests, so youll be able to see what is required via the `scripts`, and `e2e-tests` directories. (It's still in a draft, but I should have it done today).

In regards to how we fetch that test data. 
So using subscan is definitely the best way as they have a great indexer. For example, lets work through `/accounts/{accountId}/vesting-info?at=<blockHeight>`

First you can check the vesting module events, like [this](https://karura.subscan.io/event?address=&module=vesting&event=all&startDate=&endDate=&startBlock=&endBlock=&timeType=date&version=2011). You can click on the first dispatched event, `vesting(claimed)`. That will have the [accountId](https://karura.subscan.io/extrinsic/0xd4284bd8102e020abd654c239e33d034b2b8300dbbdbcb9cb4a255d53a856dd0?event=1206965-21) in the event, and you can use sidecar to query that account for its vesting info, and store that response for your testing. Similar for staking. It's all about knowing and searching for the events needed, then extracting the necessary data for sending the query. Now what can be difficult is getting the data necessary for a specific runtime, this can be kinda manual. I use `/runtime/spec?at=<blockHeight>` to confirm a block heights runtime, then will use that blockHeight as a query parameter. (Not all events will have taken place at every runtime, as something like vesting can be few and far between sometimes). 

For something like crowdloans, and auctions. It just takes a little investigating, such as looking for events when the auction ended then using the information to know what range of blocks it was going on for, then querying between that range. The best way to deduce what events you are looking for would be to 

1. use the module tab to find what module is fitting, than too look through the events and see which will be applicable to helping you find the information you need. 

2. Look directly at the source code, and what events are available per module. 

Note: I'll also be adding more to the e2e-tests as time goes by, so you should see support for pallets, and more soon. 

Hope that helps, feel free to keep asking questions if you're ever stuck

**Comment by veydecapia** [comment]
Wow! Thank you for the detailed explanation, so on your example on vesting.
I've first connected with public karura network.
<pre><font color="#3465A4">SAS:</font>
<font color="#06989A">  📦 LOG:</font>
    <font color="#4E9A06"> ✅ LEVEL: </font><font color="#D3D7CF">&quot;info&quot;</font>
    <font color="#4E9A06"> ✅ JSON: </font><font color="#D3D7CF">false</font>
    <font color="#4E9A06"> ✅ FILTER_RPC: </font><font color="#D3D7CF">false</font>
    <font color="#4E9A06"> ✅ STRIP_ANSI: </font><font color="#D3D7CF">false</font>
<font color="#06989A">  📦 SUBSTRATE:</font>
    <font color="#4E9A06"> ✅ WS_URL: </font><font color="#D3D7CF">&quot;wss://karura.api.onfinality.io/public-ws&quot;</font>
    <font color="#4E9A06"> ✅ TYPES_BUNDLE: </font><font color="#D3D7CF">undefined</font>
    <font color="#4E9A06"> ✅ TYPES_CHAIN: </font><font color="#D3D7CF">undefined</font>
    <font color="#4E9A06"> ✅ TYPES_SPEC: </font><font color="#D3D7CF">undefined</font>
    <font color="#4E9A06"> ✅ TYPES: </font><font color="#D3D7CF">undefined</font>
<font color="#06989A">  📦 EXPRESS:</font>
    <font color="#4E9A06"> ✅ BIND_HOST: </font><font color="#D3D7CF">&quot;127.0.0.1&quot;</font>
    <font color="#4E9A06"> ✅ PORT: </font><font color="#D3D7CF">8080</font>
2021-12-29 22:48:13 <font color="#4E9A06">info</font>: Version: 11.3.5


2021-12-29 22:48:33 <font color="#C4A000">warn</font>: REGISTRY: Unknown signed extensions SetEvmOrigin found, treating them as no-effect
2021-12-29 22:48:35 <font color="#4E9A06">info</font>: Connected to chain Karura on the karura client at wss://karura.api.onfinality.io/public-ws
2021-12-29 22:48:35 <font color="#4E9A06">info</font>: Listening on http://127.0.0.1:8080/
2021-12-29 22:48:35 <font color="#4E9A06">info</font>: Check the root endpoint (http://127.0.0.1:8080/) to see the available endpoints for the current node
2021-12-29 22:51:13 <font color="#C4A000">warn</font>: GET /accounts/pzNr6hfLnBLrePzPF9QJpy1nbRVb3jhSHuGSAdDtetWHHfh/vesting-info?at=1206965 404 2ms
</pre>

Then, I have tested the particular [accountId](https://karura.subscan.io/extrinsic/0xd4284bd8102e020abd654c239e33d034b2b8300dbbdbcb9cb4a255d53a856dd0?event=1206965-21) and it is not getting the account. I'm not sure what I missed.

<pre><font color="#4E9A06"><b>harvey@harvey-MS-7C92</b></font>:<font color="#3465A4"><b>~</b></font>$ curl -s http://0.0.0.0:8080/accounts/pzNr6hfLnBLrePzPF9QJpy1nbRVb3jhSHuGSAdDtetWHHfh/vesting-info?at=1206965
&lt;!DOCTYPE html&gt;
&lt;html lang=&quot;en&quot;&gt;
&lt;head&gt;
&lt;meta charset=&quot;utf-8&quot;&gt;
&lt;title&gt;Error&lt;/title&gt;
&lt;/head&gt;
&lt;body&gt;
&lt;pre&gt;Cannot GET /accounts/pzNr6hfLnBLrePzPF9QJpy1nbRVb3jhSHuGSAdDtetWHHfh/vesting-info&lt;/pre&gt;
&lt;/body&gt;
&lt;/html&gt;
</pre>

**Comment by TarikGul** [comment]
Ahh okay, its because Karura doesnt have vesting activated in the chains controller config [here](https://github.com/paritytech/substrate-api-sidecar/blob/master/src/chains-config/karuraControllers.ts#L11). I believe if you want to have it working just add in `'AccountsVestingInfo'`

**Comment by TarikGul** [comment]
I guess vesting was a bad example as their api seems to be a little different than what sidecar offers: 

That's the error I get when I added it to the config, and queried accounts.
```
{
	"code": 500,
	"message": "historicApi.query.vesting.vesting is not a function",
	"stack": "TypeError: historicApi.query.vesting.vesting is not a function\n    at AccountsVestingInfoService.fetchAccountVestingInfo (/Users/tarik/Desktop/parity/substrate-api-sidecar/build/src/services/accounts/AccountsVestingInfoService.js:21:39)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n    at async AccountsVestingInfoController.getAccountVestingInfo (/Users/tarik/Desktop/parity/substrate-api-sidecar/build/src/controllers/accounts/AccountsVestingInfoController.js:41:62)\n    at async /Users/tarik/Desktop/parity/substrate-api-sidecar/build/src/controllers/AbstractController.js:168:9",
	"level": "error"
}
```

**Comment by veydecapia** [comment]
Okay, in my understanding, vesting info is not supported at the moment for Karura network.
As you have suggested, the workaround is to manipulate the chains controller. Will definitely try this and let you know if works. Thank you!

**Comment by veydecapia** [comment]
Just have found out that parachain  is also not yet supported.

<pre><font color="#4E9A06"><b>harvey@harvey-MS-7C92</b></font>:<font color="#3465A4"><b>~</b></font>$ curl -s http://0.0.0.0:8080/paras/auctions/current?at=840000 | jq
<b>{</b>
<b>  </b><font color="#3465A4"><b>&quot;code&quot;</b></font><b>: </b>500<b>,</b>
<b>  </b><font color="#3465A4"><b>&quot;message&quot;</b></font><b>: </b><font color="#4E9A06">&quot;Parachains are not yet supported on this network.&quot;</font><b>,</b>
<b>  </b><font color="#3465A4"><b>&quot;stack&quot;</b></font><b>: </b><font color="#4E9A06">&quot;Error: Parachains are not yet supported on this network.\n    at ParasController.checkParasModule (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/@substrate/api-sidecar/build/src/controllers/paras/ParasController.js:46:23)\n    at ParasController.getAuctionsCurrent (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/@substrate/api-sidecar/build/src/controllers/paras/ParasController.js:40:18)\n    at /home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/@substrate/api-sidecar/build/src/controllers/AbstractController.js:168:15\n    at Layer.handle [as handle_request] (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/layer.js:95:5)\n    at next (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/route.js:137:13)\n    at Route.dispatch (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/route.js:112:3)\n    at Layer.handle [as handle_request] (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/layer.js:95:5)\n    at /home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/index.js:281:22\n    at Function.process_params (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/index.js:341:12)\n    at next (/home/harvey/.npm/_npx/8bf7ace223ab1c10/node_modules/express/lib/router/index.js:275:10)&quot;</font><b>,</b>
<b>  </b><font color="#3465A4"><b>&quot;level&quot;</b></font><b>: </b><font color="#4E9A06">&quot;error&quot;</font>
<b>}</b>
</pre>

**Comment by veydecapia** [comment]
Hey @TarikGul 

Hope you don't mind if I had another follow up question.
So I'm seeing a couple of endpoints you have. And for each one you have comment like the version [here](https://github.com/paritytech/substrate-api-sidecar/blob/3136399458ec3954d1305ce2c7311e873bd687a5/e2e-tests/endpoints/polkadot/accounts/balance-info/index.ts#L13) (e.g. // v0).
I assume that is the spec version, and you are also testing different spec version for each folder test case you have.
So my question is how do you fetch one in the subscan?

Tried filtering though using the SpecVersion combobox, then after clicking Filter button it is not filtering correctly based on the SpecVersion.

![image](https://user-images.githubusercontent.com/6094567/147740125-3d3c3dc5-302f-4103-b279-2eaa362f342e.png)

**Comment by TarikGul** [comment]
Hey @veydecapia, sorry for the delayed response, I was on break. 

That being said, yes you are absolutely correct, those are comments noting which spec version each call is fetching from. So each endpoint is treated a little different when it comes to testing their specVersion. For example, kusama account-balance-info has changed I believe 3 times since Kusama's first runtime, therefore we make sure to check every single runtime we can since the beginning, in order to make sure we are compatible with them all, it also uses the api's `query` methods which is more likely subject to change than the blanket rpc calls.  Whereas runtime>code just uses the core rpc call `getStorage` which most likely wont change, but also the response is quite large, so I dont see any benefit in storing all the really large files in sidecar. 

So for finding all the runtimes, I actually did it manually/automated via a script. I scraped each chain finding all the runtimes, and then making sure to find responses that were adequate for the e2e-tests. I am actually not familiar with using the specVersion tab in subscan, and from a quick try I didn't get the responses I was expecting either.
